File: py/src/syntagma/telemetry.py
# Public API — most functions called by CLI and MCP entry points.
"""Syntagma telemetry — PostHog (product analytics) + Sentry (error monitoring).

Architecture: epiccounty/reports/telemetry_sentry_posthog_architecture_2026-04-29.md
  - PostHog: single project, ``product=syntagma`` property
  - Sentry:  SENTRY_DSN_SYNTAGMA secret
  - Consent: opt-out (on by default), on/off only
  - PII: strictly forbidden — enum values only, no code / file paths / query text

Consent flow:
  - New users:     ``syntagma install`` wizard sets consent explicitly
  - Existing users / MCP clients: first binary invocation auto-enables
    telemetry and prints a one-time opt-out notice
"""

from __future__ import annotations

import hashlib
import importlib.metadata
import os
import platform
import subprocess
import sys
import time
import uuid
from enum import Enum
from pathlib import Path
from typing import Optional

# ── Enum-gated event values (no free strings) ────────────────────────────────


class Tool(str, Enum):
    SearchKnowledge = "search_knowledge"
    GetEntity = "get_entity"
    GetNeighbors = "get_neighbors"
    FindPath = "find_path"
    AnalyzeCode = "analyze_code"
    SuggestRefactorings = "suggest_refactorings"


class Command(str, Enum):
    Install = "install"
    Build = "build"
    Analyze = "analyze"
    Infer = "infer"
    Explore = "explore"
    Api = "api"
    Mcp = "mcp"
    Service = "service"
    Telemetry = "telemetry"


class FailureClass(str, Enum):
    GraphLoadError = "graph_load_error"
    EmbeddingError = "embedding_error"
    DatabaseError = "database_error"
    PermissionDenied = "permission_denied"
    NetworkError = "network_error"
    Timeout = "timeout"
    Unknown = "unknown"

    def should_capture_sentry(self) -> bool:
        return self in (FailureClass.GraphLoadError, FailureClass.DatabaseError)


class ResultSizeBucket(str, Enum):
    Empty = "empty"
    Lt5 = "<5"
    Lt20 = "<20"
    Gte20 = ">=20"

    @classmethod
    def from_count(cls, n: int) -> "ResultSizeBucket":
        if n == 0:
            return cls.Empty
        if n < 5:
            return cls.Lt5
        if n < 20:
            return cls.Lt20
        return cls.Gte20


# ── Consent ───────────────────────────────────────────────────────────────────

_SYNTAGMA_HOME = Path(os.getenv("SYNTAGMA_HOME", Path.home() / ".syntagma"))
_CONSENT_FILE = _SYNTAGMA_HOME / "telemetry-consent"
_INSTALL_ID_FILE = _SYNTAGMA_HOME / "install-id"


def read_consent_raw() -> Optional[bool]:
    """Return True/False if set, None if unset."""
    try:
        val = _CONSENT_FILE.read_text(encoding="utf-8").strip()
        if val == "off":
            return False
        if val in ("on", "true", "1"):
            return True
    except OSError:
        pass
    return None


def read_consent() -> bool:
    val = read_consent_raw()
    return val if val is not None else False


def write_consent(enabled: bool) -> None:
    _SYNTAGMA_HOME.mkdir(parents=True, exist_ok=True)
    _CONSENT_FILE.write_text("on" if enabled else "off", encoding="utf-8")


def ensure_consent_or_set_default() -> None:
    """Auto-enable on first run; print a one-time opt-out notice."""
    if read_consent_raw() is None:
        write_consent(True)
        print("[syntagma] Telemetry enabled (anonymous install ID).", file=sys.stderr)
        print("[syntagma] To opt out: syntagma telemetry off", file=sys.stderr)
        print(
            "[syntagma] Details: https://github.com/epicsagas/Syntagma#telemetry",
            file=sys.stderr,
        )


def prompt_consent_interactive() -> bool:
    """Print consent box and prompt. Returns True = enabled."""
    print()
    print("  ┌─ Telemetry ──────────────────────────────────────────────────────────┐")
    print("  │ Syntagma collects anonymous usage data to improve detection quality. │")
    print("  │                                                                      │")
    print("  │  What we send:    tool name, duration, outcome, version, OS          │")
    print("  │  What we never:   code content, file paths, search queries           │")
    print("  │  Identifier:      random install ID (not linked to you or machine)   │")
    print("  │  Opt out anytime: syntagma telemetry off                             │")
    print("  └──────────────────────────────────────────────────────────────────────┘")
    print()
    try:
        ans = input("  Enable telemetry? [Y/n]: ").strip().lower()
    except (EOFError, KeyboardInterrupt):
        ans = ""
    enabled = ans not in ("n", "no")
    return enabled


# ── Install ID ───────────────────────────────────────────────────────────────


def _load_or_create_install_id() -> str:
    try:
        val = _INSTALL_ID_FILE.read_text(encoding="utf-8").strip()
        if val:
            return val
    except OSError:
        pass
    new_id = str(uuid.uuid4())
    try:
        _SYNTAGMA_HOME.mkdir(parents=True, exist_ok=True)
        _INSTALL_ID_FILE.write_text(new_id, encoding="utf-8")
    except OSError:
        pass
    return new_id


# ── Keys (injected at build time via hatchling hook) ─────────────────────────


def _posthog_key() -> Optional[str]:
    try:
        from syntagma._keys import POSTHOG_KEY  # type: ignore[import]

        return POSTHOG_KEY or None
    except ImportError:
        pass
    return None


def _sentry_dsn() -> Optional[str]:
    try:
        from syntagma._keys import SENTRY_DSN_SYNTAGMA  # type: ignore[import]

        return SENTRY_DSN_SYNTAGMA or None
    except ImportError:
        pass
    return None


def _version() -> str:
    try:
        return importlib.metadata.version("syntagma")
    except importlib.metadata.PackageNotFoundError:
        return "unknown"


def _os_name() -> str:
    s = platform.system().lower()
    if s == "darwin":
        return "macos"
    if s == "windows":
        return "windows"
    return "linux"


# ── PostHog transport (curl, fire-and-forget) ─────────────────────────────────

_POSTHOG_HOST = "app.posthog.com"
_POSTHOG_PORT = 443


def _posthog_send(payload: str) -> None:
    key = _posthog_key()
    if not key:
        return
    body = f'{{"api_key":"{key}","batch":[{payload}]}}'
    url = f"https://{_POSTHOG_HOST}/batch/"
    try:
        subprocess.Popen(
            [
                "curl", "-s", "--max-time", "5",
                "-X", "POST",
                "-H", "Content-Type: application/json",
                "-d", body,
                url,
            ],
            stdin=subprocess.DEVNULL,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
        )
    except OSError:
        pass


def _track(event: str, props: dict) -> None:
    if not read_consent():
        return
    install_id = _load_or_create_install_id()
    ts = time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())
    base_props = {
        "product": "syntagma",
        "version": _version(),
        "os": _os_name(),
        **props,
    }
    props_json = ", ".join(
        f'"{k}": "{v}"' for k, v in base_props.items()
    )
    payload = (
        f'{{"event": "{event}", "distinct_id": "{install_id}", '
        f'"timestamp": "{ts}", "properties": {{{props_json}}}}}'
    )
    _posthog_send(payload)


# ── Sentry transport ──────────────────────────────────────────────────────────


def _parse_sentry_dsn(dsn: str) -> tuple[str, str, str]:
    """Return (host, project_path, public_key) or ('', '', '')."""
    try:
        without_scheme = dsn.removeprefix("https://")
        at = without_scheme.index("@")
        public_key = without_scheme[:at]
        rest = without_scheme[at + 1:]
        slash = rest.index("/")
        host = rest[:slash]
        project = rest[slash:]
        return host, f"/api{project}envelope/", public_key
    except (ValueError, IndexError):
        return "", "", ""


def _sentry_send(message: str, failure_class: str) -> None:
    dsn = _sentry_dsn()
    if not dsn:
        return
    host, path, public_key = _parse_sentry_dsn(dsn)
    if not host:
        return
    ver = _version()
    event_id = hashlib.md5(f"{message}{time.time()}".encode()).hexdigest()
    ts = time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())
    envelope = (
        "{}\n"
        '{"type":"event"}\n'
        f'{{"event_id":"{event_id}","timestamp":"{ts}",'
        f'"message":"{message}",'
        f'"level":"error",'
        f'"tags":{{"failure_class":"{failure_class}","product":"syntagma"}},'
        f'"release":"{ver}","platform":"python"}}\n'
    )
    auth = (
        f"Sentry sentry_version=7,"
        f"sentry_key={public_key},"
        f"sentry_client=syntagma/{ver}"
    )
    try:
        subprocess.Popen(
            [
                "curl", "-s", "--max-time", "5",
                "-X", "POST",
                "-H", f"X-Sentry-Auth: {auth}",
                "-H", "Content-Type: application/x-sentry-envelope",
                "-d", envelope,
                f"https://{host}{path}",
            ],
            stdin=subprocess.DEVNULL,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
        )
    except OSError:
        pass


# ── Typed event helpers ───────────────────────────────────────────────────────


def track_session_started() -> None:
    _track("session_started", {})


def track_command_invoked(command: Command) -> None:
    _track("command_invoked", {"command": command.value})


def track_command_completed(command: Command, duration_ms: int) -> None:
    _track("command_completed", {
        "command": command.value,
        "duration_ms": str(duration_ms),
    })


def track_command_failed(command: Command, failure: FailureClass) -> None:
    _track("command_failed", {
        "command": command.value,
        "failure_class": failure.value,
    })
    if failure.should_capture_sentry():
        _sentry_send(f"command_failed:{command.value}", failure.value)


def track_tool_called(tool: Tool) -> None:
    _track("tool_called", {"tool": tool.value})


def track_tool_completed(tool: Tool, duration_ms: int, result_size: ResultSizeBucket) -> None:
    _track("tool_completed", {
        "tool": tool.value,
        "duration_ms": str(duration_ms),
        "result_size": result_size.value,
    })


def track_tool_failed(tool: Tool, failure: FailureClass) -> None:
    _track("tool_failed", {
        "tool": tool.value,
        "failure_class": failure.value,
    })
    if failure.should_capture_sentry():
        _sentry_send(f"tool_failed:{tool.value}", failure.value)


def track_install_completed(tool_count: int) -> None:
    _track("install_completed", {"tool_count": str(tool_count)})


# ── CLI (syntagma telemetry on|off|status) ────────────────────────────────────


def run_cli(action: str = "status") -> int:
    action = action.strip().lower()
    if action == "on":
        write_consent(True)
        print("[syntagma] Telemetry enabled.")
        print("[syntagma] To opt out: syntagma telemetry off")
        return 0
    if action == "off":
        write_consent(False)
        print("[syntagma] Telemetry disabled.")
        print("[syntagma] To re-enable: syntagma telemetry on")
        return 0
    # status (default)
    val = read_consent_raw()
    if val is None:
        state = "unset (will auto-enable on next command)"
    elif val:
        state = "enabled"
    else:
        state = "disabled"
    print(f"[syntagma] Telemetry: {state}")
    print(f"[syntagma] Consent file: {_CONSENT_FILE}")
    print(f"[syntagma] Install ID:   {_INSTALL_ID_FILE}")
    return 0

File: src/adapters/telemetry.rs
use std::time::Duration;
use std::{fs, io::Write};

use tokio::sync::mpsc;

#[derive(Debug, Clone, Copy)]
pub enum Tool {
    SearchKnowledge,
    GetEntity,
    GetNeighbors,
    FindPath,
    AnalyzeCode,
    SuggestRefactorings,
}

impl Tool {
    fn as_str(self) -> &'static str {
        match self {
            Self::SearchKnowledge => "search_knowledge",
            Self::GetEntity => "get_entity",
            Self::GetNeighbors => "get_neighbors",
            Self::FindPath => "find_path",
            Self::AnalyzeCode => "analyze_code",
            Self::SuggestRefactorings => "suggest_refactorings",
        }
    }
}

#[derive(Debug, Clone, Copy)]
pub enum Command {
    Install,
    Build,
    Analyze,
    Infer,
    Explore,
    Api,
    Mcp,
    Service,
    Telemetry,
}

impl Command {
    fn as_str(self) -> &'static str {
        match self {
            Self::Install => "install",
            Self::Build => "build",
            Self::Analyze => "analyze",
            Self::Infer => "infer",
            Self::Explore => "explore",
            Self::Api => "api",
            Self::Mcp => "mcp",
            Self::Service => "service",
            Self::Telemetry => "telemetry",
        }
    }
}

#[derive(Debug, Clone, Copy)]
pub enum FailureClass {
    GraphLoadError,
    EmbeddingError,
    DatabaseError,
    PermissionDenied,
    NetworkError,
    Timeout,
    Unknown,
}

impl FailureClass {
    fn as_str(self) -> &'static str {
        match self {
            Self::GraphLoadError => "graph_load_error",
            Self::EmbeddingError => "embedding_error",
            Self::DatabaseError => "database_error",
            Self::PermissionDenied => "permission_denied",
            Self::NetworkError => "network_error",
            Self::Timeout => "timeout",
            Self::Unknown => "unknown",
        }
    }
}

#[derive(Debug, Clone, Copy)]
pub enum ResultSizeBucket {
    Empty,
    Lt5,
    Lt20,
    Gte20,
}

impl ResultSizeBucket {
    fn as_str(self) -> &'static str {
        match self {
            Self::Empty => "empty",
            Self::Lt5 => "<5",
            Self::Lt20 => "<20",
            Self::Gte20 => ">=20",
        }
    }
}

impl From<usize> for ResultSizeBucket {
    fn from(value: usize) -> Self {
        if value == 0 {
            Self::Empty
        } else if value < 5 {
            Self::Lt5
        } else if value < 20 {
            Self::Lt20
        } else {
            Self::Gte20
        }
    }
}

#[derive(Clone)]
pub struct Telemetry {
    enabled: bool,
    sender: Option<mpsc::UnboundedSender<TelemetryEvent>>,
}

#[derive(Debug, Clone)]
struct TelemetryEvent {
    event: String,
    properties: serde_json::Value,
}

impl Telemetry {
    pub fn new(
        enabled: bool,
        posthog_api_key: String,
        posthog_host: String,
        sentry_dsn: String,
    ) -> Self {
        if !enabled {
            return Self {
                enabled: false,
                sender: None,
            };
        }
        let (tx, mut rx) = mpsc::unbounded_channel::<TelemetryEvent>();
        let ph_key = posthog_api_key.clone();
        let ph_host = posthog_host.clone();
        let sentry = sentry_dsn.clone();
        std::thread::spawn(move || {
            let rt = tokio::runtime::Builder::new_current_thread()
                .enable_all()
                .build()
                .expect("telemetry runtime");
            rt.block_on(async move {
                let client = reqwest::Client::new();
                while let Some(ev) = rx.recv().await {
                    let _ = send_with_retry(&client, &ph_key, &ph_host, &sentry, &ev).await;
                }
            });
        });
        Self {
            enabled,
            sender: Some(tx),
        }
    }

    pub fn disabled() -> Self {
        Self::new(false, String::new(), String::new(), String::new())
    }

    pub fn is_enabled(&self) -> bool {
        self.enabled
    }

    pub fn track_event(&self, event: &'static str, properties: serde_json::Value) {
        if !self.enabled || !read_consent() {
            return;
        }
        let Some(tx) = &self.sender else {
            return;
        };
        let _ = tx.send(TelemetryEvent {
            event: event.to_owned(),
            properties: serde_json::json!({
                "install_id": load_or_create_install_id(),
                "payload": properties,
            }),
        });
    }
}

fn consent_file() -> std::path::PathBuf {
    crate::adapters::paths::episteme_home().join("telemetry-consent")
}

fn install_id_file() -> std::path::PathBuf {
    crate::adapters::paths::episteme_home().join("install-id")
}

pub fn read_consent_raw() -> Option<bool> {
    let path = consent_file();
    let text = fs::read_to_string(path).ok()?;
    match text.trim() {
        "on" | "true" | "1" => Some(true),
        "off" | "false" | "0" => Some(false),
        _ => None,
    }
}

pub fn read_consent() -> bool {
    read_consent_raw().unwrap_or(true)
}

pub fn write_consent(enabled: bool) -> Result<(), String> {
    let home = crate::adapters::paths::episteme_home();
    fs::create_dir_all(&home).map_err(|e| e.to_string())?;
    fs::write(consent_file(), if enabled { "on" } else { "off" }).map_err(|e| e.to_string())
}

pub fn ensure_consent_or_set_default() -> Result<(), String> {
    if read_consent_raw().is_none() {
        write_consent(true)?;
        eprintln!("[episteme] Telemetry enabled (anonymous install ID).");
        eprintln!("[episteme] To opt out: epis telemetry off");
        eprintln!("[episteme] Details: https://github.com/epicsagas/Episteme#telemetry");
    }
    Ok(())
}

pub fn prompt_consent_interactive() -> bool {
    println!();
    println!("  ┌─ Telemetry ──────────────────────────────────────────────────────────┐");
    println!("  │ Episteme collects anonymous usage data to improve detection quality. │");
    println!("  │                                                                      │");
    println!("  │  What we send:    tool name, duration, outcome, version, OS          │");
    println!("  │  What we never:   code content, file paths, search queries           │");
    println!("  │  Identifier:      random install ID (not linked to you or machine)   │");
    println!("  │  Opt out anytime: epis telemetry off                             │");
    println!("  └──────────────────────────────────────────────────────────────────────┘");
    println!();
    print!("  Enable telemetry? [Y/n]: ");
    let _ = std::io::stdout().flush();
    let mut line = String::new();
    if std::io::stdin().read_line(&mut line).is_err() {
        return true;
    }
    let ans = line.trim().to_lowercase();
    !(ans == "n" || ans == "no")
}

pub fn load_or_create_install_id() -> String {
    let path = install_id_file();
    if let Ok(s) = fs::read_to_string(&path) {
        let v = s.trim();
        if !v.is_empty() {
            return v.to_owned();
        }
    }
    let id = uuid::Uuid::new_v4().to_string();
    if let Some(parent) = path.parent() {
        let _ = fs::create_dir_all(parent);
    }
    if let Ok(mut f) = fs::File::create(&path) {
        let _ = f.write_all(id.as_bytes());
    }
    id
}

async fn send_with_retry(
    client: &reqwest::Client,
    posthog_api_key: &str,
    posthog_host: &str,
    sentry_dsn: &str,
    ev: &TelemetryEvent,
) -> Result<(), ()> {
    const MAX_RETRIES: usize = 3;
    let mut attempt = 0usize;
    loop {
        let mut ok = true;
        if !posthog_api_key.is_empty() {
            let posthog_url = format!("{}/capture/", posthog_host.trim_end_matches('/'));
            let res = client
                .post(posthog_url)
                .json(&serde_json::json!({
                    "api_key": posthog_api_key,
                    "event": ev.event,
                    "distinct_id": "episteme-server",
                    "properties": ev.properties,
                }))
                .send()
                .await;
            ok &= res
                .as_ref()
                .map(|r| r.status().is_success())
                .unwrap_or(false);
        }
        if !sentry_dsn.is_empty() {
            let res = client
                .post(sentry_dsn)
                .json(&serde_json::json!({
                    "message": ev.event,
                    "level": "info",
                    "extra": ev.properties,
                }))
                .send()
                .await;
            ok &= res
                .as_ref()
                .map(|r| r.status().is_success())
                .unwrap_or(false);
        }
        if ok {
            return Ok(());
        }
        attempt += 1;
        if attempt >= MAX_RETRIES {
            tracing::warn!(event = ev.event, "telemetry delivery failed after retries");
            return Err(());
        }
        tokio::time::sleep(Duration::from_millis(250 * (1 << attempt))).await;
    }
}

fn telemetry_client() -> Option<Telemetry> {
    let cfg = crate::adapters::config::EpistemeConfig::load().ok()?;
    if !cfg.telemetry_enabled {
        return None;
    }
    Some(Telemetry::new(
        true,
        cfg.posthog_api_key,
        cfg.posthog_host,
        cfg.sentry_dsn,
    ))
}

pub fn track_session_started() {
    if let Some(t) = telemetry_client() {
        t.track_event("session_started", serde_json::json!({}));
    }
}

pub fn track_command_invoked(command: Command) {
    if let Some(t) = telemetry_client() {
        t.track_event(
            "command_invoked",
            serde_json::json!({"command": command.as_str()}),
        );
    }
}

pub fn track_command_completed(command: Command, duration_ms: u128) {
    if let Some(t) = telemetry_client() {
        t.track_event(
            "command_completed",
            serde_json::json!({"command": command.as_str(), "duration_ms": duration_ms.to_string()}),
        );
    }
}

pub fn track_command_failed(command: Command, failure: FailureClass) {
    if let Some(t) = telemetry_client() {
        t.track_event(
            "command_failed",
            serde_json::json!({"command": command.as_str(), "failure_class": failure.as_str()}),
        );
    }
}

pub fn track_tool_called(tool: Tool) {
    if let Some(t) = telemetry_client() {
        t.track_event("tool_called", serde_json::json!({"tool": tool.as_str()}));
    }
}

pub fn track_tool_completed(tool: Tool, duration_ms: u128, result_size: ResultSizeBucket) {
    if let Some(t) = telemetry_client() {
        t.track_event(
            "tool_completed",
            serde_json::json!({
                "tool": tool.as_str(),
                "duration_ms": duration_ms.to_string(),
                "result_size": result_size.as_str()
            }),
        );
    }
}

pub fn track_tool_failed(tool: Tool, failure: FailureClass) {
    if let Some(t) = telemetry_client() {
        t.track_event(
            "tool_failed",
            serde_json::json!({"tool": tool.as_str(), "failure_class": failure.as_str()}),
        );
    }
}

pub fn track_install_completed(tool_count: usize) {
    if let Some(t) = telemetry_client() {
        t.track_event(
            "install_completed",
            serde_json::json!({"tool_count": tool_count.to_string()}),
        );
    }
}
